### PR TITLE
feat: redesign update price and transfer item on gallery

### LINKED ISSUE #4543: Gallery Item Desktop - integrate list, transfer, and change price
## Ref
- https://github.com/kodadot/nft-gallery/issues/4231

### PR DESCRIPTION
**Pull request by preschian**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [x] Feature
- [x] Refactoring

## Context

- [x] Closes #4543

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=DY4SQF2iD456tH89aQtz5wv1EV3BbSW8wKKuMcwbmXaj1pM)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemRelist.vue [hunk 1]
```diff
@@ -1,62 +1,106 @@
...
+            props.nftId,
+            String(meta)
+          ),
+        ]
+        break
+
+      case 'snek':
+      case 'bsx':
+        cb = getApiCall(api, urlPrefix.value, 'LIST')
+        arg = bsxParamResolver(props.nftId, 'LIST', String(meta))
```
Comment (Jarsen136) [inline components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemRelist.vue:92]: `Interaction` would help

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemTransfer.vue [hunk 2]
```diff
@@ -1,35 +1,110 @@
...
+    }
+
+    let cb, arg
+
+    switch (urlPrefix.value) {
+      case 'rmrk':
+        cb = api.tx.system.remark
+        arg = [
+          createInteraction(
+            'SEND' as JustInteraction,
```
Comment (Jarsen136) [inline components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemTransfer.vue:81]: `JustInteraction.SEND` or `Interaction.SEND` could help

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemRelist.vue [hunk 3]
```diff
@@ -1,62 +1,106 @@
...
+  nftId: string
+  nftPrice: string
+}>()
 
-function toggleActive() {
-  active.value = !active.value
-}
+const active = ref(false)
+const price = ref()
+const isListed = computed(() => Number(props.nftPrice))
```
Comment (Jarsen136) [inline components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemRelist.vue:57]: It should be a Boolean?

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemRelist.vue [hunk 4]
```diff
@@ -1,62 +1,106 @@
...
+    }
+
+    let cb, arg
+
+    switch (urlPrefix.value) {
+      case 'rmrk':
+        cb = api.tx.system.remark
+        arg = [
+          createInteraction(
+            'LIST' as JustInteraction,
```
Comment (Jarsen136) [inline components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemRelist.vue:81]: Interaction would help

File: composables/useMetaTransaction.ts [hunk 5]
```diff
@@ -28,6 +29,7 @@ function useMetaTransaction() {
     onError?: () => void
   ): Promise<void> => {
     try {
+      isLoading.value = true
```
Comment (Jarsen136) [inline composables/useMetaTransaction.ts:32]: How about using `initTransactionLoader` outside  the composables?

**Comment by roiLeo** [comment]
Hey! 
I see you're writting some actions logic and I was wondering if it was really a good idea to have a switch case instead of a if statement?
Shouldn't it be moved to a file outside this component?

File: composables/useMetaTransaction.ts (hunk 5 above)
Comment (Jarsen136) [reply composables/useMetaTransaction.ts:32]: I just switch to your branch to try to reproduce it. However, I remove this line `isLoading.value = true` and add `initTransactionLoader()` before `howAboutToExecute(xxx)`. The loader would show up. How about try again?

image

image

**Review by vikiival — CHANGES_REQUESTED** [review]
Hello, please resolve code climate issues

I find `useGalleryItemAction` really hard to read

**Comment by vikiival** [comment]
Last things:
- how would you handle new chain ?

**Comment by vikiival** [comment]
Tested list and send 

https://deploy-preview-4652--koda-nuxt.netlify.app/rmrk/gallery/16143102-8cc1b91e899d9be40e-2CENTS-BREAKFAST_IN_SUBWORK-0000000000000008

Looks like it works

File: components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemRelist.vue [hunk 6]
```diff
@@ -1,62 +1,73 @@
...
+            @click.native="updatePrice" />
+        </template>
+
+        <template #content>
+          <div>
+            <input
+              v-model="price"
+              type="number"
+              :placeholder="
+                isListed ? 'Your New Price' : 'Your Listing Price'
```
Comment (vikiival) [inline components/gallery/GalleryItemAction/GalleryItemActionType/GalleryItemRelist.vue:22]: add to translations

**Review by vikiival — APPROVED** [review]
Code LGTM.

I would like to invite other pair of eyes for this

File: components/gallery/GalleryItemAction/GalleryItemActionType/useGalleryItemAction.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    dangerMessage(err)
+    return
+  }
+}
+
+function constructTransactionSend(item: ActionSend, api, executeTransaction) {
+  const { id, item: token } = tokenIdToRoute(item.tokenId)
+  checkBeforeSend(item)
+
+  if (item.urlPrefix === 'rmrk') {
```
Comment (Jarsen136) [inline components/gallery/GalleryItemAction/GalleryItemActionType/useGalleryItemAction.ts:87]: using `switch & case` pattern to replace if/else would be better.

**Review by Jarsen136 — COMMENTED** [review]
except for some new translations words, otherwise lgtm

File: components/gallery/GalleryItemAction/GalleryItemActionType/useGalleryItemAction.ts (hunk 7 above)
Comment (Jarsen136) [reply components/gallery/GalleryItemAction/GalleryItemActionType/useGalleryItemAction.ts:87]: IMO there is one main difference to `if/else`, `switch & case` would be easy to extend more option cases. 
By the way, object literal would also be a good option, that was an implement of `Strategy Pattern`.

Comment (vikiival) [reply components/gallery/GalleryItemAction/GalleryItemActionType/useGalleryItemAction.ts:87]: I think we should wrap it into SDK 
- https://github.com/kodadot/packages/issues/89

**Review by Jarsen136 — APPROVED** [review]

**Comment by vikiival** [comment]
@preschian please add translations and we can move it

**Pull request merged**